internal/jsondb/db.go

```diff
@@ -0,0 +1,106 @@
+// Package jsondb implements a simple database of JSON documents, backed by the
+// file system.
+//
+// It supports two operations: Read() and Write(). Their signatures mirror
+// those of json.Unmarshal() and json.Marshal():
+//
+//     err := db.Write("my-string", "octopus")
+//
+//     var v string
+//     exists, err := db.Read("my-string", &v)
+//
+// The JSON documents are stored in a directory, in the form name.json (name as
+// passed to Read() and Write()). Thus, names may only contain characters that
+// may appear in filenames.
+
+package jsondb
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+type JSONDatabase struct {
+	dir  string
+	perm os.FileMode
+}
+
+// Create a new JSONDatabase in `dir`. Each document that is saved to it will
+// have a file mode of `perm`.
+func New(dir string, perm os.FileMode) *JSONDatabase {
+	return &JSONDatabase{dir, perm}
+}
+
+// Reads the value at `name`. `document` must deserializable from JSON. Returns
+// false if a document with `name` does not exist or an error occured.
+func (db *JSONDatabase) Read(name string, document interface{}) (bool, error) {
```

Comment:
Hm, using bool as a special error feels odd.

Reply:
Yeah I was unsure here. I went with what I thought is a common Go conventions, especially the `map` type, which I think is most similar in concept: 
```golang
// maps
value, ok := m["foo"]
// type assertion
i, ok := v.(int)
// os.LookupEnv
v, ok := os.LookupEnv("CACHE_DIRECTORY")
```

The alternative would be to return a `os.ErrNotExist` error and make callers check for that. That would leak the underlying implmentation, no?

Another alternative would be to define a custom error type `jsondb.ErrorNotExist`. (I think I prefer this one if you really don't like the `bool` return.)

Reply:
Ah, I see it now. Fair enough, thanks!

Reply:
I've (hopefully) clarified the doc string.